rfc/openfed__prerequisite.md

```diff
@@ -0,0 +1,408 @@
+---
+title: "@openfed__prerequisite directive to conduct behaviour before another operation."
+author: David Stutt
+---
+
+# Overview
+The `@openfed__prerequisite` directive (henceforth simply "prerequisite") intends to be a flexible and powerful addition
+to GraphQL Federation.
+When included in a schema, the directive will always perform its defined behaviour before the rest of the operation.
+This would allow the user to:
+- Validate an input that references an entity.
+- Validate an input against another operation.
+
+Potential idea and discussion point:
+- Inject input through the response of an exterior query.
+
+# Definition
+This is the proposed definition of the directive (functionality is discussed in the appropriate section):
+```graphql
+directive @openfed__prerequisite(
+  resolveEntity: openfed__ResolveEntityInput
+  resolveQuery: openfed__ResolveQueryInput
+) on ARGUMENT_DEFINITION
+
+input openfed__ResolveEntityInput {
+  typeName: String!
+  fields: openfed__InputSet!
+}
+
+input openfed__ResolveQueryInput {
+  query: String!
+}
+
+scalar openfed__InputSet
+```
+
+# Functionality
+
+The `prerequisite` directive (currently) defines two optional arguments.
+The arguments are mutually exclusive; and although they are both optional (limitation of GraphQL), one (and only one) 
+argument _must_ be defined.
+Failure to define an argument, or defining both, will result in a composition error.
+
+## 1. resolveEntity argument
+The `resolveEntity` argument requires two inputs:
+- `typeName`: The `__typename` of the entity to be validated.
+- `fields`: The key field(s) that form the primary key.
+
+### The typeName input
+The `typeName` input corresponds to the `__typename` of an entity that defines an entity resolver in the same subgraph
+that defines the directive.
+
+Composition will fail if the type referenced by the `typeName` input is not an Object that defines a valid `@key`
+directive with the `resolvable` argument either omitted or explicitly set to true.
+
+### The fields input
+The `fields` input is type `openfed__InputSet!` (henceforth simply `InputSet`).
```

Comment:
I am not convinced the name of this input is best. Some other potential ideas:

- `MappedFieldSet`
- `MapSet`
- `InputToFieldSet`

Reply:
I think that `InputToFieldSet` is too verbose. I personally like `MappedFieldSet`

Reply:
Or you can make it clear that it's specific for `openfed__ResolveEntityInput` by calling it `ResolveFieldSet`